compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -511,6 +516,31 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         };
         obligations.extend(traits::predicates_for_generics(cause.clone(), self.param_env, bounds));
 
+        // If we're producing obligations for an operator trait, and it has an `Output`
+        // associated type, then we add an `Output=expected` obligation.
+        if let (true, Expectation::ExpectHasType(output_ty)) = (is_op, expected) {
+            let opt_output_assoc_item = tcx.associated_items(trait_def_id).find_by_name_and_kind(
+                tcx,
+                Ident::from_str("Output"),
+                AssocKind::Type,
+                trait_def_id,
+            );
+            if let Some(output_assoc_item) = opt_output_assoc_item {
+                obligations.push(traits::Obligation::new(
+                    cause.clone(),
+                    self.param_env,
+                    ty::Binder::dummy(ty::PredicateKind::Projection(ProjectionPredicate {
```

Comment:
My only worry is that registering this projection obligation will cause inference to break in certain places, since it asserts type equality as opposed to coercibility.